src/lib/crypto/snarky_tests/snarky_tests.ml

```diff
@@ -191,6 +191,88 @@ let generate_witness () =
 
   ()
 
+module As_prover_circuits = struct
+  let ( -- ) i j =
+    let rec aux n acc = if n < i then acc else aux (n - 1) (n :: acc) in
+    aux j []
+
+  let input_typ = Impl.Typ.tuple3 Impl.Field.typ Impl.Field.typ Impl.Field.typ
+
+  let return_typ = Impl.Typ.unit
+
+  let main as_prover vars
+      ((b1, b2, b3) : Impl.Field.t * Impl.Field.t * Impl.Field.t) () =
+    let abc = Impl.Field.(b1 + b2 + b3) in
+    if as_prover then
+      Impl.as_prover (fun _ ->
+          let f acc e =
+            let var = Snarky_backendless.Cvar.(Var e) in
+            (* let u = Impl.Field.Constant.(random ()) in *)
+            let v = Impl.As_prover.read_var var in
```

Comment:
I'm not sure this is a good way to write the test, because it assumes that the index of b1, b2, b3 will be 1, 2, 3, which is not going to be true very soon (once we get rid of the R1CS trick, the index of these vars should start at 0). 

Two solution to this:

* assert that the index behind `b1` is 1, so that when we do the R1CS change we can fail early here and fix the test
* use the index behind `b1` as your start offset

random notes:

you could also test `read_var (b1 :> Snarky_backendless.Cvar.t)` here, like so maybe:

```ocaml
let x1 = read_var (b1 :> Snarky_backendless.Cvar.t) in
let x2 = read b1 in
assert(Field.Constant.Equal x1 x2)
```

there's also some other functions you can test: `let x1 = read_var (Field.var_indices b1)`. Not sure why we have so many functions : ))

Reply:
I will go with the first, I prefer to have an explicit number of indices to check, so that when we have the two implementations, they will be consistent.
How does `var_indices` work? it seems to return a list.

Reply:
also, where is the `:>` in `read_var (b1 :> Snarky_backendless.Cvar.t)` implemented?

Reply:
@fabrizio-m `:>` is the coercion operator in OCaml (see https://v2.ocaml.org/manual/expr.html#ss%3Aexpr-coercions)